Upload component during the create flow

jhadvig opened the pull request:
Adding option to add to project any component that we expose to the users. For this I've created a separate tab in the 'Add to project' page, together with 'From Docker Image'.
Component could be added in a multiple ways:
1. Browse trough the file on users local machine by hitting the "Browse" button right to the input field. 
2. Paste component description into ace editor beneath the input field.
3. Drag and Drop the file from users local machine to the area of the input field or ace editor.
JSON and YAML format are supported.

Also upon componet creation a check is being done to find out whether a component with the same name doesn't exist in current project. If it exists it will be replaced** with the new one. If it doesn't it will be created.
** The way how replacing is working is that we have to delete the old component and create the new one instead. The old component cant be updated with the new one since the new one is missing data like timestamps, revision number, generation , ...
This is a bit unfortunate approach but wasnt sure if extending data of the old component with the data from the new one was better since by extending the fields from the old component would be preserved.

Had to update the oscFileInput scope with an `fileExtension` and `dropArea` variable. Originally wanted to have whole filename, but realize that more valid information then filename(which could be anything) is the file extention. The `dropArea` variable points to a area which should respond to Drag'n'Drop action.

Also for me unknown reason for me when redirecting upon template creation to fromTemplate page and upon imageStream creation to fromImage page  with `$location.path` (create.html#206 && create.html#210) I'm redirected to project list. The path is though right cause when I replace it with `location.assign` method (which is not what we want to use) Im getting the desired redirection.

bindata.go is missing, will provide after review.

Actual screenshot:
![screenshot](https://cloud.githubusercontent.com/assets/1668218/13921557/b3accdaa-ef78-11e5-9cf8-96c85eaf1004.png)
![screenshot-1](https://cloud.githubusercontent.com/assets/1668218/13923150/ac00da0e-ef7f-11e5-92fe-ff75e03d2ed6.png)

@jwforres @spadgett PTAL

jwforres commented:
@jhadvig did you try using PATCH instead of POST for the replace?  i'm not sure with the CLI does today for the replace @fabianofranz ?

liggitt commented:
triggering off the file extension seems fragile to me. can we not do content detection?

jwforres commented:
@jhadvig you are already doing content detection if the fileExtension is not there (I assume for the direct paste into the editor case?), why not always do content detection?  seems like an unnecessary extra step to validate that the file they upload with extension yml actually parses as YAML.  I think what you are doing with attempting the JSON parse first and then the YAML is probably the right order, since the JSON parser is native to the browser and so hopefully its been optimized.

jwforres commented:
the Cancel button doesn't seem to be working

jwforres commented:
Would it be possible to change the ace editor's mode based on what content you have populated it with?  @spadgett do you know if you can change the editor's mode after its created?

looking at JSON when in YAML mode isn't great

jwforres commented:
I used examples/ha/openshift-ha.yaml and got this:

![empty_reason](https://cloud.githubusercontent.com/assets/1467629/13960669/a4ab69c6-f02f-11e5-9318-d86ab9d6cce0.png)

jhadvig commented:
@jwforres yes I guess we can skip the file extension check and just do the content detection.
Regarding the ace editor mode, it looks like its possible. Will try to change in upon the aceChanged event and set based on content detection.

jhadvig commented:
@jwforres I didnt try the PATCH cause I don't think its the right approach/action, cause some fields from the old resource could be left and unwanted, after patching with the new one.
@fabianofranz could you please share info about what CLI does today for the replace action. Thanks !

fabianofranz commented:
@jhadvig @jwforres `oc replace` does a PUT to update the object, unless it's explicitly set to force with `oc replace --force` then in deletes and creates again (DELETE/POST). `oc patch` does a PATCH, but it's targeted to specifically update some fields instead of the entire object, e.g. `oc patch ... -p '{"spec":{"unschedulable":true}}'`.

jwforres commented on assets/app/scripts/controllers/create.js:
```diff
@@ -94,6 +100,132 @@ angular.module('openshiftConsole')
+          $scope.alerts['parsing'] = {
+            type: "error",
+            message: "An error occurred during parsing editors content.",
+            details: "Reason: " + $filter('getErrorDetails')(e)
```
getErrorDetails was written to get errors from API responses, the error here is a parse exception, i think you want to take a look at what @spadgett does in the YAML editor as far as handling these exceptions, cause you care about the snippet that the parse error occurred on

jwforres commented on assets/app/scripts/controllers/create.js:
```diff
@@ -94,6 +100,132 @@ angular.module('openshiftConsole')
+          $scope.alerts['parsing'] = {
+            type: "error",
+            message: "An error occurred during parsing editors content.",
+            details: "Reason: " + $filter('getErrorDetails')(e)
```
same thing except a JSON parse exception is probably going to be a different format than the yaml parse exception, just try putting in bad JSON and see what the exception looks like

jwforres commented:
I think there are two cases that oc create handles that aren't being correctly handled here:
1) kind List, when you see Kind List you are supposed to create each of the things in the List
2) kind SomeOtherKindList, like ImageStreamList, same thing iterate over it and create the things in the list

liggitt commented:
I think the pattern is

* `kind` = `List` or `*List`
* `items` = Array

jhadvig commented:
@jwforres I've updated the PR. Added comments to the functions, that should clarify the purpose.
Regarding https://github.com/openshift/origin/pull/8173#issuecomment-199912739 the problem is that we dont have that one it in proper yaml format. Try putting it into http://yamltojson.com/ and convert. It will fail.
PTAL

jwforres commented:
In the case where you specify a template, I don't like that we actually add the template itself in the project without asking you.  Maybe when someone selects a template we should show something like:

You are adding a template, you can: 
- [x] Process the template - this will create an instance of the things defined in the template, you will have an opportunity to fill out any parameters the template defines
- [ ] Add template - this will make the template available to others who have access to this project
 
@spadgett thoughts?

jwforres commented:
and by default you would have process checked and add unchecked, since i think that is going to be more common

liggitt commented:
> and by default you would have process checked and add unchecked, since i think that is going to be more common

+1

spadgett commented:
@jwforres Sounds good.

jwforres commented:
@jhadvig also can you re-open this PR against jwforres:console_1.3 branch?  There are enough changes in the branch at this point that testing this PR which is against origin may not catch everything.

jwforres commented:
also dont include bindata.go in the commit when you open it against console_1.3

jhadvig commented:
@jwforres @spadgett I've prepared couple of proposals* for the mentioned issue. All of them will pop up a modal that is notifying user that he is adding a Template and that he has to pick up an action. I've picked a modal cause I didnt come up with a better way how to model the picking the next action upon the Template creation (didnt like the idea of scanning the ace editor after every change and dynamically add bunch of buttons, like Save, Process, etc..., or checkboxes).
*following comments present the proposals

jhadvig commented:
**Proposal 1**
![screenshot-7](https://cloud.githubusercontent.com/assets/1668218/14318861/1adc5412-fc0f-11e5-8f43-fa8cef17c379.png)

jhadvig commented:
**Proposal 2**
![screenshot-8](https://cloud.githubusercontent.com/assets/1668218/14318893/41e6658e-fc0f-11e5-8761-0ada65d6553c.png)
![screenshot-9](https://cloud.githubusercontent.com/assets/1668218/14318897/4311642c-fc0f-11e5-9df9-2fc90da2aaef.png)

jhadvig commented:
**Proposal 3**
In this one I've also added third option and thats *Save and Process*.
Under the selectbox will be the description of the action.
![screenshot-3](https://cloud.githubusercontent.com/assets/1668218/14318925/5e954e02-fc0f-11e5-8413-6ab4f065c81a.png)
![screenshot-4](https://cloud.githubusercontent.com/assets/1668218/14318927/5fbcdda4-fc0f-11e5-98e8-2b859438e2d0.png)

jhadvig commented:
@jwforres @spadgett 
Also the *Add* option should be check by default I think. The reason is that I as a user will be pretty upset when I will be writing the template from scratch into the ace editor and something will go wrong on the process page and my template wont be saved. Thoughts ?  
Thats why I also came up the a standalone option that combines those two(in case we will go with radio or selectbox).

spadgett commented:
@jhadvig I like proposal 1. Maybe use `class="help-block"` below the checkbox like we use in other places.

spadgett commented:
@jhadvig Can you rebase your work on https://github.com/jwforres/origin/tree/console_1.3? We've been adding 1.3 content there.

spadgett commented on assets/app/scripts/controllers/create.js:
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
+        } catch (e) {
+          return;
+        }
+      }
```
We should debounce these checks so we're not parsing and reparsing on every keystroke.

https://lodash.com/docs#debounce

spadgett commented on assets/app/scripts/controllers/create.js:
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
+    $scope.aceLoaded = function(editor) {
+      $scope.editor.session = editor.getSession();
+      $scope.editor.session.setOption('tabSize', 2);
+      $scope.editor.session.setOption('useSoftTabs', true);
```
Since this isn't used in the view anywhere, let's make it a local var to the controller.

```javascript
var editorSession;
$scope.aceLoaded = function(editor) {
      editorSession = editor.getSession();

...
```

spadgett commented on assets/app/scripts/controllers/create.js:
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
+      $scope.editor.empty = false;
+      if (!$scope.editor.content) {
+        $scope.editor.empty = true;
+      }
```
I'd just check `!editor.content` in the view and remove `editor.empty`.

spadgett commented on assets/app/scripts/controllers/create.js:
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
+            type: "error",
+            message: "An error occurred during parsing editors content.",
+            details: "Reason: " + e.message
+          };
```
Would be nice to give additional error details like

https://github.com/openshift/origin/blob/master/assets/app/scripts/controllers/modals/editModal.js#L69
https://github.com/openshift/origin/blob/master/assets/app/views/modals/edit-resource.html#L14-L16

spadgett commented on assets/app/scripts/controllers/create.js:
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
+      $scope.resourceKind = resource.kind;
+      
+      if ($scope.resourceKind.endsWith("List")) {
+        $scope.resourceList = resource.items;
```
```
$scope.resourceName = '';
```

?

spadgett commented on assets/app/scripts/controllers/create.js:
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
+
+      var itemsProcessed = 0;
+      angular.forEach($scope.resourceList, function(item) {
+        var itemKind = item.kind;
```
Show an error if kind if missing.

jhadvig replied:
Well I though about adding and error but I wasn't sure if we wanna have the validation done client side. Then we would have to check for all the fields that have to be filled for each resources kind. Do we want it ?

jhadvig replied:
Sorry I take it back, I've realized that we need it.

spadgett commented on assets/app/scripts/controllers/create.js:
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
+        var itemKind = item.kind;
+        var itemName = item.metadata.name;
+
+        if (item.metadata.namespace && item.metadata.namespace !== $scope.projectName) {
```
These lines will result in a runtime error if metadata is missing entirely.

spadgett commented on assets/app/scripts/controllers/create.js:
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
+        }
+
+        // Check if the resource already exists. If it does, replace it spec with the new one.
+        DataService.get(APIService.kindToResource(itemKind), itemName, {namespace: $scope.projectName}, {errorNotification: false}).then(
```
Better to move this code somewhere inside `ProjectsService.get()` and use `context` from the callback.

spadgett commented on assets/app/scripts/controllers/create.js:
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
+          subPath = "overview";
+        }
+        $location.url("project/" + encodeURIComponent($scope.projectName) + "/" + subPath);
+        $scope.$evalAsync();
```
Why is this needed?

spadgett commented on assets/app/scripts/controllers/create.js:
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
+              } else {
+                createAndUpdate();
+              }
+            }
```
You should be able to move this code after the `angular.forEach`, then you don't need to track `itemsProcessed`. If you want to track if there were errors, you could set a flag that you check later.

spadgett commented on assets/app/scripts/controllers/create.js:
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
+      });
+    };
+
+    function openModal() {
```
`promptToReplaceResource`?

spadgett replied:
Or `confirmReplace`?

spadgett commented on assets/app/scripts/controllers/create.js:
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
+      var modalInstance = $uibModal.open({
+        animation: true,
+        templateUrl: 'views/modals/update-resource.html',
+        controller: 'UpdateModalController',
```
I'd give the view and controller different names. This sounds like the edit resource modal. Maybe `confirm-replace.html` and `ConfirmReplaceModalController`?

spadgett commented on assets/app/scripts/controllers/create.js:
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
+    function createAndUpdate() {
+      $scope.createdAndUpdateResources = 0;
+
+      if (!_.isEmpty($scope.updateResources)) {
```
The `_.isEmpty` check shouldn't be needed. It's handled by `angular.forEach`.

spadgett commented on assets/app/scripts/controllers/create.js:
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
+      if ($scope.createdAndUpdateResources === $scope.resourceList.length) {
+        var subPath;
+        if ($scope.resourceKind === "Template") {
+          subPath =  "create/fromtemplate?name=" + $scope.resourceName + "&namespace=" + encodeURIComponent($scope.projectName);
```
If creating a list of templates, will this start creating the last template? Probably should go back to the overview.

jhadvig replied:
`$scope.resourceKind` is holding the top level resource Kind. So in the case of TemplateList it will be redirected to overview

spadgett commented:
We should use `TaskList` for create/updating lists of things. Have a look at

https://github.com/openshift/origin/blob/master/assets/app/scripts/controllers/create/createFromImage.js#L214-L258

spadgett commented on assets/app/scripts/controllers/create.js:
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
+    // Redirect to appropriate page based on the created/updated resource. If resource is Template redirect to
+    // frotemplate page, else to the project overview.
+    function redirectIfComplete() {
+      $scope.createdAndUpdateResources++;
```
You should add all the promises to an array and use `$q.all` instead of counting.

https://docs.angularjs.org/api/ng/service/$q#all

spadgett commented on assets/app/scripts/directives/oscFileInput.js:
```diff
@@ -8,15 +8,25 @@ angular.module('openshiftConsole')
+        var dropArea = element;
+        if (scope.dropArea) {
+          dropArea = scope.dropArea;
+        }
```
```
var dropArea = scope.dropArea || element;
```

jwforres replied:
This reminds me, when I was trying this yesterday... We don't give you any
indication of what the drop target is, I dropped it outside of the editor
area and it replaced the whole page with the contents of the file.
On Apr 6, 2016 2:26 PM, "Sam Padgett" <notifications@github.com> wrote:

> In assets/app/scripts/directives/oscFileInput.js
> <https://github.com/openshift/origin/pull/8173#discussion_r58756740>:
>
> >        },
> >        templateUrl: 'views/directives/osc-file-input.html',
> >        link: function(scope, element){
> >          scope.helpID = _.uniqueId('help-');
> >          scope.supportsFileUpload = (window.File && window.FileReader && window.FileList && window.Blob);
> >          scope.uploadError = false;
> > +        var dropArea = element;
> > +        if (scope.dropArea) {
> > +          dropArea = scope.dropArea;
> > +        }
>
> var dropArea = scope.dropArea || element;
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8173/files/fe69812b62eb038054b34c55a4709deaacdbaefc#r58756740>
>

spadgett commented on assets/app/scripts/directives/oscFileInput.js:
```diff
@@ -31,6 +41,28 @@ angular.module('openshiftConsole')
+        });
+        $(dropArea).on('dragenter', function(e) {
+            e.preventDefault();
+            e.stopPropagation();
```
Indentation

spadgett commented on assets/app/scripts/directives/oscFileInput.js:
```diff
@@ -31,6 +41,28 @@ angular.module('openshiftConsole')
           reader.readAsText(file);
         });
+        $('.btn-file').on('click', function() {
+          delete scope.file;
```
Is this triggered when the button is disabled (since it's just a span)? What happens if I cancel the dialog?

spadgett commented on assets/app/scripts/directives/oscFileInput.js:
```diff
@@ -31,6 +41,28 @@ angular.module('openshiftConsole')
+              scope.file = e.originalEvent.dataTransfer.files[0];
+              $(element).trigger('change');
+            }
+          }
```
Maybe

```javascript
var files = _.get(e, 'originalEvent.dataTransferFiles', [])
if (files.length > 0) {
  ...
}
```

jhadvig replied:
:+1:

spadgett commented on assets/app/scripts/directives/oscFileInput.js:
```diff
@@ -31,6 +41,28 @@ angular.module('openshiftConsole')
+          if (e.originalEvent.dataTransfer){
+            if (e.originalEvent.dataTransfer.files.length > 0) {
+              scope.file = e.originalEvent.dataTransfer.files[0];
+              $(element).trigger('change');
```
Is this to trigger the handler above? I'd rather pull that out into its own method and call it directly.

jhadvig replied:
You mean that the `$(element).trigger('change');` will be in a standalone method,which we can then call to trigger?

spadgett commented on assets/app/views/create.html:
```diff
@@ -10,156 +10,199 @@
-                              <a href="" ng-click="filter.tag = tag">{{tag}}</a>
-                            </li>
-                          </ul>
+                <h1>Add Components to your Project</h1>
```
Some push back on the new wording in #7961 :/

spadgett replied:
Let's use the header "Add to Project" until we have API approval to make a change (see #7961)

spadgett commented on assets/app/views/create.html:
```diff
@@ -10,156 +10,199 @@
-                          </ul>
+                <h1>Add Components to your Project</h1>
+
+                <uib-tabset justified="true">
```
Did we decide if we want justified or not?

jhadvig replied:
Well there was a discussion and I know that @rhamilto suggested to not have it justified. But then @ajacobs21e supplied the wireframes for the new create page and there it was justified, so I've figured out I will leave it justified and lets discuss it on the PR, here :)

spadgett commented on assets/app/views/create.html:
```diff
@@ -10,156 +10,199 @@
+                              ng-click="create()"
+                              ng-disabled="editor.empty"
+                              class="btn btn-primary btn-lg">
+                              Add
```
Should probably be "Create" to match `oc create`. Also per Patternfly guidelines:

https://www.patternfly.org/styles/terminology-and-wording/#action-labels

spadgett commented on assets/app/views/create.html:
```diff
@@ -10,156 +10,199 @@
+
+                  <uib-tab active="selectedTab.fromDockerImage">
+                    <uib-tab-heading>From Docker Image</uib-tab-heading>
+                  </uib-tab>
```
Let's remove the last tab from your PR. I'll add it in mine.

spadgett commented on assets/app/views/modals/update-resource.html:
```diff
@@ -0,0 +1,26 @@
+<div class="modal-resource-delete">
+  <div class="modal-body">
+    <div ng-if="resourceList.length === 1">
+      <h1>{{resourceKind | humanize}} '<strong>{{resourceName}}</strong>' already exists</h1>
```
This filter was renamed `humanizeKind`.

spadgett replied:
For when you rebase.

jhadvig replied:
:+1:

spadgett commented on assets/app/views/modals/update-resource.html:
```diff
@@ -0,0 +1,26 @@
+    </div>
+    
+    <div ng-if="resourceList.length > 1">
+      <h1>Some of {{resourceKind | humanize}} items already exists</h1>
```
What if there are multiple resources with different kinds? I'd just say "Some items already exist"

jhadvig replied:
There can be only one top level `Kind`, in this particular case it will be List. Listed will be only those items from the List that already exist, whatever type they are.

spadgett replied:
So it says "Some of List items already exists"? In that case, I'd still just say "Some items already exist"

jhadvig replied:
No I meant, it will say: `Following components already exist:`

spadgett replied:
Let's say items for now until we get agreement on terminology.

jhadvig replied:
:+1:

spadgett commented on assets/app/views/modals/update-resource.html:
```diff
@@ -0,0 +1,26 @@
+    
+    <div ng-if="resourceList.length > 1">
+      <h1>Some of {{resourceKind | humanize}} items already exists</h1>
+      <p>Following components already exist:</p>
```
"The following items already exist:"

?

jhadvig replied:
Hmmm, will merge the h1 and p into one header saying:
"Following List items already exists"

spadgett commented on assets/app/views/modals/update-resource.html:
```diff
@@ -0,0 +1,26 @@
+        <div ng-repeat="resource in updateResources">
+          <dt>{{resource.kind}}</dt>
+          <dd>{{resource.metadata.name}}</dd>
+        </div>
```
I think this will work:

```html
<dl class="dl-horizontal">
  <dt ng-repeat-start="resource in updateResources">...</dt>
  <dd ng-repeat-end>...</dd>
</dl>
```

spadgett commented on assets/app/scripts/controllers/create.js:
```diff
@@ -94,6 +107,213 @@ angular.module('openshiftConsole')
+        height: Math.floor(window.innerHeight * 0.50) + 'px'
+      }, 30, function() {
+        editor.resize();
+      });
```
Should we handle window resizes?

spadgett commented:
We should think about breaking up the create controller because it's getting big. Maybe create from file is a directive included in the page.

jhadvig commented:
@spadgett rebased && added the bindata

spadgett commented on assets/app/scripts/controllers/newfromtemplate.js:
```diff
@@ -188,17 +196,38 @@ angular.module('openshiftConsole')
+                details: "Not found in cache."
+              };
+            $scope.templateForm.$setValidity("",false);
+            return;
```
We should navigate to the error page here, or improve the error message. It currently looks like this.

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/14885920/18380c38-0d1c-11e6-8cc7-b44feb87a70d.png)

jhadvig replied:
Yeah, redirecting to error page would be better

spadgett commented:
Seeing some runtime errors on add to project after picking a source builder:

![openshift_web_console](https://cloud.githubusercontent.com/assets/1167259/14885977/5a7a86d4-0d1c-11e6-9ac3-2e0765b25045.png)

spadgett commented on assets/app/scripts/directives/fromFile.js:
```diff
@@ -0,0 +1,397 @@
+            if ($scope.resourceKind === "Template") {
+              $scope.templateOptions = {
+                process: true,
+                add: true
```
Missed this on earlier review, but @jwforres and @liggitt suggested add be false by default.

jhadvig commented:
@spadgett all three fixed. Thanks !

jhadvig commented:
The thing I wondering about is if we dont wanna have the 'From catalog' and 'From file' tab headers justified. Or not justifying was the final decision ? 
@spadgett 
cc'ing @jwforres @rhamilto

spadgett commented:
> The thing I wondering about is if we dont wanna have the 'From catalog' and 'From file' tab headers justified. 

I prefer unjustified for what it's worth.

spadgett commented:
@jhadvig The projects dropdown stops working for me once I drag a file. Do you see this as well?

spadgett commented:
Add to Project -> From File -> Drop a file -> Click the project name in the breadcrumb -> Click the projects downdrop

jhadvig commented:
@spadgett yes I can, but its really weird !?! Will take a look whats causing the problem

jhadvig commented:
hmm ... I can even reproduce the issue when redirecting to 'Add to project' and then go back to the overview

spadgett commented on assets/app/scripts/directives/oscFileInput.js:
```diff
@@ -32,6 +141,19 @@ angular.module('openshiftConsole')
+
+        scope.$on('$destroy', function(){
+          $(dropZoneSelectorName).off();
+          $(document).off('drop').off('dragenter').off('dragover').off('dragleave');
```
Can you give these namespaces so we don't accidentally remove other drag/drop event handlers that might be registered? Maybe `"drop.oscFileInput"` when you call `on()`. Then `$(document).off('.oscFileInput')` should be enough to unregister all the different event types (untesed).

http://api.jquery.com/on/
http://api.jquery.com/off/

jhadvig commented:
@spadgett thanks so much for the great suggestion ! I've updated the PR

spadgett commented:
LGTM, @jwforres suggested a fork AMI so QE can test IE

jhadvig commented:
@spadgett ok will create one right away

spadgett commented:
[merge]

jhadvig commented:
@spadgett after your PR get merger I had to rebase my (not fully sure why though...)
But its rebased now :)

spadgett commented:
[merge]

jhadvig commented:
@spadgett thanks so much for assistance on this !!!
:joy:
